rewriter/Minitest.cc

```diff
@@ -452,19 +634,114 @@ ast::ExpressionPtr runSingle(core::MutableContext ctx, bool isClass, ast::Send *
+                                          flattenDescribeBody(rhs.deepCopy()));
+
+        // Create a filtered module version that contains only includable methods (let, etc.)
+        auto moduleName = makeSharedExamplesModuleConstantLocal(ctx, arg.loc(), argString);
+
+        // Filter the RHS to include only methods that should be includable
+        ast::ClassDef::RHS_store filteredModuleRhs;
+        auto flattenedRhs = flattenDescribeBody(move(rhs));
+
+        // Add extend T::Helpers at the beginning
+        auto tHelpers = ast::MK::EmptyTree();
+        tHelpers = ast::MK::UnresolvedConstant(arg.loc(), move(tHelpers), core::Names::Constants::T());
+        tHelpers = ast::MK::UnresolvedConstant(arg.loc(), move(tHelpers), core::Names::Constants::Helpers());
+        auto extendStatement = ast::MK::Send1(send->loc, ast::MK::Self(send->loc), core::Names::extend(), send->funLoc,
+                                              std::move(tHelpers));
+        filteredModuleRhs.emplace_back(std::move(extendStatement));
+
+        // Add requires_ancestor { RSpec::Core::ExampleGroup }
+        if (!send->recv.isSelfReference() && isRSpec(send->recv)) {
+            // Add requires_ancestor when we have explicit RSpec receiver (RSpec context)
+            auto rspecExampleGroup = ast::MK::EmptyTree();
+            rspecExampleGroup =
+                ast::MK::UnresolvedConstant(arg.loc(), move(rspecExampleGroup), core::Names::Constants::RSpec());
+            rspecExampleGroup =
+                ast::MK::UnresolvedConstant(arg.loc(), move(rspecExampleGroup), core::Names::Constants::Core());
+            rspecExampleGroup =
+                ast::MK::UnresolvedConstant(arg.loc(), move(rspecExampleGroup), core::Names::Constants::ExampleGroup());
+
+            // Create a block that returns the RSpec::Core::ExampleGroup constant
+            auto blockBody = std::move(rspecExampleGroup);
+            ast::MethodDef::ARGS_store blockArgs;
+            auto block = ast::MK::Block(send->loc, std::move(blockBody), std::move(blockArgs));
+
+            auto requiresAncestorStatement = ast::MK::Send0Block(
+                send->loc, ast::MK::Self(send->loc), core::Names::requiresAncestor(), send->loc, std::move(block));
+            filteredModuleRhs.emplace_back(std::move(requiresAncestorStatement));
+        }
+
+        for (auto &expr : flattenedRhs) {
+            // Only include method definitions (let, subject, etc.) in the module
+            // Skip describe blocks and other non-includable constructs
+            if (auto method = ast::cast_tree<ast::MethodDef>(expr)) {
+                filteredModuleRhs.emplace_back(expr.deepCopy());
+            }
+        }
+
+        ast::ClassDef::ANCESTORS_store moduleAncestors;
+        auto moduleResult = ast::MK::Module(send->loc, declLoc, std::move(moduleName), std::move(moduleAncestors),
+                                            std::move(filteredModuleRhs));
```

Comment:
This seems weird to me. This is almost certainly not what the RSpec runtime is doing.

Why is it not possible for the Minitest.cc rewriter to model what the RSpec runtime would have done? Does RSpec create either a module, or a class? What ancestors does it give to the class or module it creates?

The current approach is not great, because it doubles the amount of code inside a shared_examples block. Sorbet tries very hard to not allocate more AST nodes than required, which is one of the things that keeps its performance good.

It also means that things like errors inside the duplicated methods will be reported twice, leaving users confused why there are duplicate errors.

Reply:
@jez this was pretty tricky for me... the reason I did this is because a shared example/context can also include tests, in the same way a describe block can – so it has to be a class for that reason because that's how we model the ability to use `it 'works' { ... }` syntax in `describe` blocks. However, we need to be able to `include` shared contexts – so I have two versions:
1. A module, that skips rewriting the `it` blocks (only available on the class), that can be included
2. A class, that includes the `it` blocks, that is not included

Do you have a better way to do this?

Reply:
Support for `it` does not fundamentally require classes:

https://sorbet.run/#%23%20typed%3A%20true%0A%0Amodule%20Foo%0A%20%20def%20something%20%3D%20nil%0A%0A%20%20it%20'test'%20do%0A%20%20%20%20something%0A%20%20end%0Aend

Reply:
@jez , I was able to make this work by having it be a module only, and using the new "magic" thing with requires ancestor to indicate that these modules require `RSpec::Core::ExampleGroup` ancestor, which makes it able to use `create` methods, and `puts` and `raise` without `Kernel`, so I think we're good here 🎉 

Commit is e99e4824b